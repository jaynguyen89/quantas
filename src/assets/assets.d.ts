declare module '*.png' {
    const path: string;
    export default path;
}

declare module 'blob-polyfill';
