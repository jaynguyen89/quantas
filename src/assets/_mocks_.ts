import { IApiHotel, IHotel } from 'src/commons/interfaces';
import { CancellationType, ImageType, PromotionType, RatingType } from 'src/commons/enums';

export const mockedApiHotelData: {
    results: Array<IApiHotel>,
} = {
    results: [
        {
            id: '123',
            property: {
                propertyId: '111',
                title: 'First Hotel',
                address: ['1 What Street', 'Where 3333'],
                previewImage: {
                    url: 'https://photo1.png',
                    caption: 'Photo1',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 3,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo1',
                    type: 'MEMBER',
                },
                name: 'Promo1',
                displayPrice: {
                    amount: 111,
                    currency: 'AU',
                },
                savings: {
                    amount: 11,
                    currency: 'AU',
                },
                cancellationOption: {
                    cancellationType: 'NOT_REFUNDABLE',
                },
            },
        },
        {
            id: '456',
            property: {
                propertyId: '222',
                title: 'Second Hotel',
                address: ['2 Which Way', 'Anywhere 2222'],
                previewImage: {
                    url: 'https://photo2.png',
                    caption: 'Photo2',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 4,
                    ratingType: 'star',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo2',
                    type: 'MEMBER',
                },
                name: 'Promo2',
                displayPrice: {
                    amount: 222,
                    currency: 'US',
                },
                savings: null,
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
        {
            id: '789',
            property: {
                propertyId: '333',
                title: 'Third Hotel',
                address: ['3 Where Road', 'Somewhere 1111', 'Australia'],
                previewImage: {
                    url: 'https://photo3.png',
                    caption: 'Photo3',
                    imageType: 'PRIMARY',
                },
                rating: {
                    ratingValue: 5,
                    ratingType: 'self',
                },
            },
            offer: {
                promotion: {
                    title: 'Promo3',
                    type: 'MEMBER',
                },
                name: 'Promo3',
                displayPrice: {
                    amount: 333,
                    currency: 'VND',
                },
                savings: {
                    amount: 33,
                    currency: 'VND',
                },
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION',
                },
            },
        },
    ],
};

export const mockedHotelData: Array<IHotel> = [
    {
        id: '123',
        name: 'First Hotel',
        address: '1 What Street, Where 3333',
        image: {
            url: 'https://photo1.png',
            caption: 'Photo1',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 3,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo1',
            price: {
                value: 111,
                currency: 'AU',
            },
            promotion: {
                title: 'Promo1',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 11,
                currency: 'AU',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '456',
        name: 'Second Hotel',
        address: '2 Which Way, Where 2222',
        image: {
            url: 'https://photo2.png',
            caption: 'Photo2',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 4,
            type: RatingType.STAR,
        },
        pricing: {
            name: 'Promo2',
            price: {
                value: 222,
                currency: 'US',
            },
            promotion: {
                title: 'Promo2',
                type: PromotionType.MEMBER,
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
    {
        id: '789',
        name: 'Third Hotel',
        address: '3 Where Road, Somewhere 1111, Australia',
        image: {
            url: 'https://photo3.png',
            caption: 'Photo3',
            type: ImageType.PRIMARY,
        },
        rating: {
            value: 5,
            type: RatingType.SELF,
        },
        pricing: {
            name: 'Promo3',
            price: {
                value: 333,
                currency: 'US',
            },
            promotion: {
                title: 'Promo3',
                type: PromotionType.MEMBER,
            },
            savings: {
                value: 33,
                currency: 'US',
            },
            cancellationType: CancellationType.FREE_CANCELLATION,
        },
    },
];