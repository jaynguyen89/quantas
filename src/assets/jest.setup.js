import { TextEncoder, TextDecoder, ReadableStream } from 'node:util';

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
global.ReadableStream = ReadableStream;