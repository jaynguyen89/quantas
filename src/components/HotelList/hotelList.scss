@use 'src/commons/variables' as vars;

.wrapper {
    margin-top: vars.$small;

    .logo {
        width: 180px;
    }

    .list-top {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        margin-top: vars.$small;
        margin-bottom: vars.$xxtiny;
        font-size: vars.$sh9;
        font-weight: vars.$regular;

        .counter {
            font-style: italic;

            & :nth-child(odd) {
                font-weight: vars.$thick;
                font-style: normal;
            }
        }
    }

    .list-body {
        width: 100%;
    }
}