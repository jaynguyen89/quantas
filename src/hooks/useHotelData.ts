import { IHotel } from 'src/commons/interfaces';
import { useEffect, useState } from 'react';
import { getApiHotelData, mapApiHotelDataToClientSchema } from 'src/utilities/utilities';

const useHotelData = () => {
    const [hotelData, setHotelData] = useState<Array<IHotel>>([]);

    useEffect(() => {
        getApiHotelData().then(data => {
            const hotelData = mapApiHotelDataToClientSchema(data);
            setHotelData(hotelData);
        });
    }, []);

    return hotelData;
};

export default useHotelData;

