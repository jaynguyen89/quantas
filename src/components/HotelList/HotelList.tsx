import React, { useEffect, useMemo, useState } from 'react';
import Container from 'src/components/Container/Container';
import Sorter from 'src/components/HotelList/Sorter/Sorter';
import HotelItem from 'src/components/HotelList/HotelItem/HotelItem';
import useHotelData from 'src/hooks/useHotelData';
import Loading from 'src/components/Loading/Loading';
import { IHotel } from 'src/commons/interfaces';
import { SortingOption } from 'src/commons/enums';
import configs from 'src/commons/configs';

import QuantasLogo from 'src/assets/qantas-logo.png';
import 'src/components/HotelList/hotelList.scss';

const HotelList = () => {
    const hotelData = useHotelData();

    const [sorter, setSorter] = useState<keyof typeof SortingOption>(configs.defaultSortOption);
    const [hotels, setHotels] = useState<Array<IHotel>>(hotelData);

    const { count, hotelText } = useMemo(() => {
        let count = 0;
        let hotelText = '';

        if (hotelData.length === 0) return { count, hotelText };

        count = hotelData.length;
        hotelText = count > 1 ? 'hotels' : 'hotel';

        return { count, hotelText };
    }, [hotelData]);



    useEffect(() => {
        if (hotelData.length === 0) return;

        const sortedData = hotelData.sort((former, latter) => {
            const comparison = former.pricing.price.value - latter.pricing.price.value;
            return SortingOption[sorter] === SortingOption.HIGH_LOW ? -1 * comparison : comparison;
        });

        setHotels([...sortedData]);
    }, [sorter, hotelData]);

    const handleChange = (option: string) =>
        setSorter(option as keyof typeof SortingOption);

    return (
        <Container>
            <div className='wrapper'>
                {(hotels.length === 0 && <Loading />) || (
                    <>
                        <img className='logo' src={QuantasLogo} alt='Quantas' />

                        <div className='list-top'>
                            <div className='counter'>
                                <span className='count'>{count}</span>
                                <span>&nbsp;{hotelText} in&nbsp;</span>
                                <span className='location'>Sydney.</span>
                            </div>

                            <Sorter
                                name='price-sorter'
                                value={sorter}
                                handleChange={handleChange}
                            />
                        </div>

                        <div className='list-body'>
                            {hotels.map(hotel => (
                                <HotelItem key={hotel.id} hotel={hotel} />
                            ))}
                        </div>
                    </>
                )}
            </div>
        </Container>
    );
};

export default HotelList;