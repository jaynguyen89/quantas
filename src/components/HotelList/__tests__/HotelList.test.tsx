import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import HotelList from 'src/components/HotelList/HotelList';
import toJson from 'enzyme-to-json';
import { render } from '@testing-library/react';
import useHotelData from 'src/hooks/useHotelData';
import { mockedHotelData } from 'src/assets/_mocks_';

configure({ adapter: new Adapter() });

jest.mock('src/hooks/useHotelData');
const mockUseHotelData = jest.mocked(useHotelData);

describe('HotelList.tsx', () => {
    it('renders correctly', () => {
        mockUseHotelData.mockReturnValue(mockedHotelData);
        const wrapper = shallow(<HotelList />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('contains the necessary elements', () => {
        mockUseHotelData.mockReturnValue(mockedHotelData);
        const wrapper = render(<HotelList />);

        // Todo: finish the test
    });
});