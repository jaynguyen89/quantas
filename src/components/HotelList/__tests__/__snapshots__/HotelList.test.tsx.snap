// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelList.tsx renders correctly 1`] = `
<Container>
  <div
    className="wrapper"
  >
    <img
      alt="Quantas"
      className="logo"
      src="test-file-stub"
    />
    <div
      className="list-top"
    >
      <div
        className="counter"
      >
        <span
          className="count"
        >
          3
        </span>
        <span>
           
          hotels
           in 
        </span>
        <span
          className="location"
        >
          Sydney.
        </span>
      </div>
      <Sorter
        handleChange={[Function]}
        name="price-sorter"
        value="HIGH_LOW"
      />
    </div>
    <div
      className="list-body"
    >
      <HotelItem
        hotel={
          Object {
            "address": "1 What Street, Where 3333",
            "id": "123",
            "image": Object {
              "caption": "Photo1",
              "type": "PRIMARY",
              "url": "https://photo1.png",
            },
            "name": "First Hotel",
            "pricing": Object {
              "cancellationType": "FREE_CANCELLATION",
              "name": "Promo1",
              "price": Object {
                "currency": "AU",
                "value": 111,
              },
              "promotion": Object {
                "title": "Promo1",
                "type": "MEMBER",
              },
              "savings": Object {
                "currency": "AU",
                "value": 11,
              },
            },
            "rating": Object {
              "type": "SELF",
              "value": 3,
            },
          }
        }
        key="123"
      />
      <HotelItem
        hotel={
          Object {
            "address": "2 Which Way, Where 2222",
            "id": "456",
            "image": Object {
              "caption": "Photo2",
              "type": "PRIMARY",
              "url": "https://photo2.png",
            },
            "name": "Second Hotel",
            "pricing": Object {
              "cancellationType": "FREE_CANCELLATION",
              "name": "Promo2",
              "price": Object {
                "currency": "US",
                "value": 222,
              },
              "promotion": Object {
                "title": "Promo2",
                "type": "MEMBER",
              },
            },
            "rating": Object {
              "type": "STAR",
              "value": 4,
            },
          }
        }
        key="456"
      />
      <HotelItem
        hotel={
          Object {
            "address": "3 Where Road, Somewhere 1111, Australia",
            "id": "789",
            "image": Object {
              "caption": "Photo3",
              "type": "PRIMARY",
              "url": "https://photo3.png",
            },
            "name": "Third Hotel",
            "pricing": Object {
              "cancellationType": "FREE_CANCELLATION",
              "name": "Promo3",
              "price": Object {
                "currency": "US",
                "value": 333,
              },
              "promotion": Object {
                "title": "Promo3",
                "type": "MEMBER",
              },
              "savings": Object {
                "currency": "US",
                "value": 33,
              },
            },
            "rating": Object {
              "type": "SELF",
              "value": 5,
            },
          }
        }
        key="789"
      />
    </div>
  </div>
</Container>
`;
